File: insight_desktop/storage/repository.py
"""The one place that runs SQL. Everything else in the app (engine, UI)
reads and writes conversation history, prompt versions, and memory facts
only through this Repository — never through raw sqlite3 calls of their
own.
"""

from __future__ import annotations

import sqlite3
import uuid
from datetime import datetime, timezone

from storage import db
from storage.models import MemoryFact, Message, PromptVersion, Session


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _new_id() -> str:
    return uuid.uuid4().hex


class Repository:
    def __init__(self, db_path: str):
        self._conn: sqlite3.Connection = db.connect(db_path)

    def close(self) -> None:
        self._conn.close()

    # ------------------------------------------------------------------
    # Sessions
    # ------------------------------------------------------------------

    def create_session(self) -> Session:
        session = Session(id=_new_id(), started_at=_now(), ended_at=None, status="active")
        self._conn.execute(
            "INSERT INTO sessions (id, started_at, ended_at, status) VALUES (?, ?, ?, ?)",
            (session.id, session.started_at, session.ended_at, session.status),
        )
        self._conn.commit()
        return session

    def end_session(self, session_id: str) -> None:
        self._conn.execute(
            "UPDATE sessions SET status = 'ended', ended_at = ? WHERE id = ?",
            (_now(), session_id),
        )
        self._conn.commit()

    def get_latest_active_session(self) -> Session | None:
        row = self._conn.execute(
            "SELECT * FROM sessions WHERE status = 'active' ORDER BY started_at DESC LIMIT 1"
        ).fetchone()
        return _row_to_session(row) if row else None

    def list_sessions(self, limit: int = 50) -> list[Session]:
        rows = self._conn.execute(
            "SELECT * FROM sessions ORDER BY started_at DESC LIMIT ?", (limit,)
        ).fetchall()
        return [_row_to_session(r) for r in rows]

    # ------------------------------------------------------------------
    # Messages
    # ------------------------------------------------------------------

    def add_message(
        self,
        session_id: str,
        role: str,
        content: str,
        source: str = "text",
        prompt_version_id: str | None = None,
        latency_ms: int | None = None,
        cancelled: bool = False,
    ) -> Message:
        message = Message(
            id=_new_id(),
            session_id=session_id,
            ts=_now(),
            role=role,
            content=content,
            source=source,
            prompt_version_id=prompt_version_id,
            latency_ms=latency_ms,
            cancelled=cancelled,
        )
        self._conn.execute(
            """INSERT INTO messages
               (id, session_id, ts, role, content, source, prompt_version_id, latency_ms, cancelled)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)""",
            (
                message.id, message.session_id, message.ts, message.role, message.content,
                message.source, message.prompt_version_id, message.latency_ms, int(message.cancelled),
            ),
        )
        self._conn.commit()
        return message

    def get_session_messages(self, session_id: str, limit: int = 500) -> list[Message]:
        rows = self._conn.execute(
            "SELECT * FROM messages WHERE session_id = ? ORDER BY ts ASC LIMIT ?",
            (session_id, limit),
        ).fetchall()
        return [_row_to_message(r) for r in rows]

    def count_session_messages(self, session_id: str) -> int:
        row = self._conn.execute(
            "SELECT COUNT(*) AS c FROM messages WHERE session_id = ?", (session_id,)
        ).fetchone()
        return int(row["c"])

    # ------------------------------------------------------------------
    # Prompt versions
    # ------------------------------------------------------------------

    def save_prompt_version(self, content: str, label: str | None = None) -> PromptVersion:
        version = PromptVersion(id=_new_id(), content=content, label=label, created_at=_now(), is_active=True)
        self._conn.execute("UPDATE prompt_versions SET is_active = 0")
        self._conn.execute(
            "INSERT INTO prompt_versions (id, content, label, created_at, is_active) VALUES (?, ?, ?, ?, 1)",
            (version.id, version.content, version.label, version.created_at),
        )
        self._conn.commit()
        return version

    def get_active_prompt_version(self) -> PromptVersion | None:
        row = self._conn.execute(
            "SELECT * FROM prompt_versions WHERE is_active = 1 ORDER BY created_at DESC LIMIT 1"
        ).fetchone()
        return _row_to_prompt_version(row) if row else None

    def activate_prompt_version(self, version_id: str) -> PromptVersion | None:
        self._conn.execute("UPDATE prompt_versions SET is_active = 0")
        self._conn.execute("UPDATE prompt_versions SET is_active = 1 WHERE id = ?", (version_id,))
        self._conn.commit()
        row = self._conn.execute("SELECT * FROM prompt_versions WHERE id = ?", (version_id,)).fetchone()
        return _row_to_prompt_version(row) if row else None

    def list_prompt_versions(self, limit: int = 50) -> list[PromptVersion]:
        rows = self._conn.execute(
            "SELECT * FROM prompt_versions ORDER BY created_at DESC LIMIT ?", (limit,)
        ).fetchall()
        return [_row_to_prompt_version(r) for r in rows]

    # ------------------------------------------------------------------
    # Memory facts
    # ------------------------------------------------------------------

    def add_memory_fact(self, text: str) -> MemoryFact:
        fact = MemoryFact(id=_new_id(), text=text, created_at=_now(), active=True)
        self._conn.execute(
            "INSERT INTO memory_facts (id, text, created_at, active) VALUES (?, ?, ?, 1)",
            (fact.id, fact.text, fact.created_at),
        )
        self._conn.commit()
        return fact

    def list_memory_facts(self, active_only: bool = True) -> list[MemoryFact]:
        if active_only:
            rows = self._conn.execute(
                "SELECT * FROM memory_facts WHERE active = 1 ORDER BY created_at ASC"
            ).fetchall()
        else:
            rows = self._conn.execute("SELECT * FROM memory_facts ORDER BY created_at ASC").fetchall()
        return [_row_to_memory_fact(r) for r in rows]

    def remove_memory_fact(self, fact_id: str) -> None:
        self._conn.execute("UPDATE memory_facts SET active = 0 WHERE id = ?", (fact_id,))
        self._conn.commit()

    def clear_all_memory_facts(self) -> None:
        self._conn.execute("UPDATE memory_facts SET active = 0")
        self._conn.commit()


# ----------------------------------------------------------------------
# Row -> dataclass helpers
# ----------------------------------------------------------------------

def _row_to_session(row: sqlite3.Row) -> Session:
    return Session(id=row["id"], started_at=row["started_at"], ended_at=row["ended_at"], status=row["status"])


def _row_to_message(row: sqlite3.Row) -> Message:
    return Message(
        id=row["id"],
        session_id=row["session_id"],
        ts=row["ts"],
        role=row["role"],
        content=row["content"],
        source=row["source"],
        prompt_version_id=row["prompt_version_id"],
        latency_ms=row["latency_ms"],
        cancelled=bool(row["cancelled"]),
    )


def _row_to_prompt_version(row: sqlite3.Row) -> PromptVersion:
    return PromptVersion(
        id=row["id"], content=row["content"], label=row["label"],
        created_at=row["created_at"], is_active=bool(row["is_active"]),
    )


def _row_to_memory_fact(row: sqlite3.Row) -> MemoryFact:
    return MemoryFact(id=row["id"], text=row["text"], created_at=row["created_at"], active=bool(row["active"]))

File: insight_ios/Packages/InsightStorage/Sources/InsightStorage/Repository.swift
import Foundation
import SQLite3

public final class Repository: @unchecked Sendable {
    private let connection: OpaquePointer

    public init(dbPath: String) throws {
        let pathURL = URL(fileURLWithPath: dbPath)
        try FileManager.default.createDirectory(
            at: pathURL.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        connection = try Database.open(at: dbPath)
    }

    init(connection: OpaquePointer) {
        self.connection = connection
    }

    public static func inMemory() throws -> Repository {
        Repository(connection: try Database.openInMemory())
    }

    deinit {
        sqlite3_close(connection)
    }

    // MARK: - Sessions

    @discardableResult
    public func createSession() -> SessionRecord {
        let session = SessionRecord(
            id: Self.newID(),
            startedAt: Self.now(),
            endedAt: nil,
            status: "active"
        )
        execute(
            "INSERT INTO sessions (id, started_at, ended_at, status) VALUES (?, ?, ?, ?)",
            bindings: [.text(session.id), .text(session.startedAt), .null, .text(session.status)]
        )
        return session
    }

    public func endSession(sessionID: String) {
        execute(
            "UPDATE sessions SET status = 'ended', ended_at = ? WHERE id = ?",
            bindings: [.text(Self.now()), .text(sessionID)]
        )
    }

    public func getLatestActiveSession() -> SessionRecord? {
        queryOne(
            "SELECT id, started_at, ended_at, status FROM sessions WHERE status = 'active' ORDER BY started_at DESC LIMIT 1",
            map: Self.mapSession
        )
    }

    public func listSessions(limit: Int = 50) -> [SessionRecord] {
        queryMany(
            "SELECT id, started_at, ended_at, status FROM sessions ORDER BY started_at DESC LIMIT ?",
            bindings: [.int(limit)],
            map: Self.mapSession
        )
    }

    // MARK: - Messages

    @discardableResult
    public func addMessage(
        sessionID: String,
        role: String,
        content: String,
        source: String = "text",
        promptVersionID: String? = nil,
        latencyMs: Int? = nil,
        cancelled: Bool = false
    ) -> MessageRecord {
        let message = MessageRecord(
            id: Self.newID(),
            sessionID: sessionID,
            timestamp: Self.now(),
            role: role,
            content: content,
            source: source,
            promptVersionID: promptVersionID,
            latencyMs: latencyMs,
            cancelled: cancelled
        )
        execute(
            """
            INSERT INTO messages
            (id, session_id, ts, role, content, source, prompt_version_id, latency_ms, cancelled)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            bindings: [
                .text(message.id),
                .text(message.sessionID),
                .text(message.timestamp),
                .text(message.role),
                .text(message.content),
                .text(message.source),
                message.promptVersionID.map(SQLValue.text) ?? .null,
                message.latencyMs.map(SQLValue.int) ?? .null,
                .int(cancelled ? 1 : 0),
            ]
        )
        return message
    }

    public func getSessionMessages(sessionID: String, limit: Int = 500) -> [MessageRecord] {
        queryMany(
            """
            SELECT id, session_id, ts, role, content, source, prompt_version_id, latency_ms, cancelled
            FROM messages WHERE session_id = ? ORDER BY ts ASC LIMIT ?
            """,
            bindings: [.text(sessionID), .int(limit)],
            map: Self.mapMessage
        )
    }

    public func countSessionMessages(sessionID: String) -> Int {
        queryOne(
            "SELECT COUNT(*) AS c FROM messages WHERE session_id = ?",
            bindings: [.text(sessionID)],
            map: { statement in
                Int(sqlite3_column_int(statement, 0))
            }
        ) ?? 0
    }

    // MARK: - Prompt versions

    @discardableResult
    public func savePromptVersion(content: String, label: String? = nil) -> PromptVersionRecord {
        let version = PromptVersionRecord(
            id: Self.newID(),
            content: content,
            label: label,
            createdAt: Self.now(),
            isActive: true
        )
        execute("UPDATE prompt_versions SET is_active = 0")
        execute(
            "INSERT INTO prompt_versions (id, content, label, created_at, is_active) VALUES (?, ?, ?, ?, 1)",
            bindings: [
                .text(version.id),
                .text(version.content),
                version.label.map(SQLValue.text) ?? .null,
                .text(version.createdAt),
            ]
        )
        return version
    }

    public func getActivePromptVersion() -> PromptVersionRecord? {
        queryOne(
            """
            SELECT id, content, label, created_at, is_active
            FROM prompt_versions WHERE is_active = 1 ORDER BY created_at DESC LIMIT 1
            """,
            map: Self.mapPromptVersion
        )
    }

    public func activatePromptVersion(versionID: String) -> PromptVersionRecord? {
        execute("UPDATE prompt_versions SET is_active = 0")
        execute(
            "UPDATE prompt_versions SET is_active = 1 WHERE id = ?",
            bindings: [.text(versionID)]
        )
        return queryOne(
            """
            SELECT id, content, label, created_at, is_active
            FROM prompt_versions WHERE id = ?
            """,
            bindings: [.text(versionID)],
            map: Self.mapPromptVersion
        )
    }

    public func listPromptVersions(limit: Int = 50) -> [PromptVersionRecord] {
        queryMany(
            """
            SELECT id, content, label, created_at, is_active
            FROM prompt_versions ORDER BY created_at DESC LIMIT ?
            """,
            bindings: [.int(limit)],
            map: Self.mapPromptVersion
        )
    }

    // MARK: - Memory facts

    @discardableResult
    public func addMemoryFact(text: String) -> MemoryFactRecord {
        let fact = MemoryFactRecord(
            id: Self.newID(),
            text: text,
            createdAt: Self.now(),
            active: true
        )
        execute(
            "INSERT INTO memory_facts (id, text, created_at, active) VALUES (?, ?, ?, 1)",
            bindings: [.text(fact.id), .text(fact.text), .text(fact.createdAt)]
        )
        return fact
    }

    public func listMemoryFacts(activeOnly: Bool = true) -> [MemoryFactRecord] {
        if activeOnly {
            return queryMany(
                "SELECT id, text, created_at, active FROM memory_facts WHERE active = 1 ORDER BY created_at ASC",
                map: Self.mapMemoryFact
            )
        }
        return queryMany(
            "SELECT id, text, created_at, active FROM memory_facts ORDER BY created_at ASC",
            map: Self.mapMemoryFact
        )
    }

    public func removeMemoryFact(factID: String) {
        execute(
            "UPDATE memory_facts SET active = 0 WHERE id = ?",
            bindings: [.text(factID)]
        )
    }

    public func clearAllMemoryFacts() {
        execute("UPDATE memory_facts SET active = 0")
    }

    // MARK: - Helpers

    private enum SQLValue {
        case text(String)
        case int(Int)
        case null
    }

    private func execute(_ sql: String, bindings: [SQLValue] = []) {
        var statement: OpaquePointer?
        guard sqlite3_prepare_v2(connection, sql, -1, &statement, nil) == SQLITE_OK else {
            fatalError("SQLite prepare failed: \(Database.lastError(from: connection))")
        }
        defer { sqlite3_finalize(statement) }

        bind(bindings, to: statement)
        guard sqlite3_step(statement) == SQLITE_DONE else {
            fatalError("SQLite execute failed: \(Database.lastError(from: connection))")
        }
    }

    private func queryOne<T>(
        _ sql: String,
        bindings: [SQLValue] = [],
        map: (OpaquePointer) -> T
    ) -> T? {
        var statement: OpaquePointer?
        guard sqlite3_prepare_v2(connection, sql, -1, &statement, nil) == SQLITE_OK else {
            fatalError("SQLite prepare failed: \(Database.lastError(from: connection))")
        }
        defer { sqlite3_finalize(statement) }

        bind(bindings, to: statement)
        guard sqlite3_step(statement) == SQLITE_ROW, let statement else {
            return nil
        }
        return map(statement)
    }

    private func queryMany<T>(
        _ sql: String,
        bindings: [SQLValue] = [],
        map: (OpaquePointer) -> T
    ) -> [T] {
        var statement: OpaquePointer?
        guard sqlite3_prepare_v2(connection, sql, -1, &statement, nil) == SQLITE_OK else {
            fatalError("SQLite prepare failed: \(Database.lastError(from: connection))")
        }
        defer { sqlite3_finalize(statement) }

        bind(bindings, to: statement)

        var rows: [T] = []
        while sqlite3_step(statement) == SQLITE_ROW, let statement {
            rows.append(map(statement))
        }
        return rows
    }

    private func bind(_ bindings: [SQLValue], to statement: OpaquePointer?) {
        for (index, value) in bindings.enumerated() {
            let position = Int32(index + 1)
            switch value {
            case let .text(text):
                sqlite3_bind_text(statement, position, text, -1, unsafeBitCast(-1, to: sqlite3_destructor_type.self))
            case let .int(number):
                sqlite3_bind_int(statement, position, Int32(number))
            case .null:
                sqlite3_bind_null(statement, position)
            }
        }
    }

    private static func now() -> String {
        ISO8601DateFormatter().string(from: Date())
    }

    private static func newID() -> String {
        UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased()
    }

    private static func mapSession(_ statement: OpaquePointer) -> SessionRecord {
        SessionRecord(
            id: columnText(statement, 0),
            startedAt: columnText(statement, 1),
            endedAt: columnOptionalText(statement, 2),
            status: columnText(statement, 3)
        )
    }

    private static func mapMessage(_ statement: OpaquePointer) -> MessageRecord {
        MessageRecord(
            id: columnText(statement, 0),
            sessionID: columnText(statement, 1),
            timestamp: columnText(statement, 2),
            role: columnText(statement, 3),
            content: columnText(statement, 4),
            source: columnText(statement, 5),
            promptVersionID: columnOptionalText(statement, 6),
            latencyMs: columnOptionalInt(statement, 7),
            cancelled: sqlite3_column_int(statement, 8) != 0
        )
    }

    private static func mapPromptVersion(_ statement: OpaquePointer) -> PromptVersionRecord {
        PromptVersionRecord(
            id: columnText(statement, 0),
            content: columnText(statement, 1),
            label: columnOptionalText(statement, 2),
            createdAt: columnText(statement, 3),
            isActive: sqlite3_column_int(statement, 4) != 0
        )
    }

    private static func mapMemoryFact(_ statement: OpaquePointer) -> MemoryFactRecord {
        MemoryFactRecord(
            id: columnText(statement, 0),
            text: columnText(statement, 1),
            createdAt: columnText(statement, 2),
            active: sqlite3_column_int(statement, 3) != 0
        )
    }

    private static func columnText(_ statement: OpaquePointer, _ index: Int32) -> String {
        String(cString: sqlite3_column_text(statement, index))
    }

    private static func columnOptionalText(_ statement: OpaquePointer, _ index: Int32) -> String? {
        guard sqlite3_column_type(statement, index) != SQLITE_NULL else {
            return nil
        }
        return columnText(statement, index)
    }

    private static func columnOptionalInt(_ statement: OpaquePointer, _ index: Int32) -> Int? {
        guard sqlite3_column_type(statement, index) != SQLITE_NULL else {
            return nil
        }
        return Int(sqlite3_column_int(statement, index))
    }
}
